bin/detect

```diff
@@ -1,10 +1,74 @@
 #!/usr/bin/env bash
 # bin/detect <build-dir>
 
+error() {
+  local c="2,999 s/^/ !     /"
+	# send all of our output to stderr
+	exec 1>&2
+
+	echo -e "\033[1;31m" # bold; red
+	echo -n " !     ERROR: "
+	# this will be fed from stdin
+  case $(uname) in
+		Darwin) sed -l "$c";; # mac/bsd sed: -l buffers on line boundaries
+		*)      sed -u "$c";; # unix/gnu sed: -u unbuffered (arbitrary) chunks of data
+	esac
+	echo -e "\033[0m" # reset style
+	exit 1
+}
+
 if [ -f $1/package.json ]; then
   echo 'Node.js'
   exit 0
 fi
 
->&2 echo 'Node.js: package.json not found in application root'
+if [[ -f "$1/.slugignore" ]] && grep -Fxq "package.json" "$1/.slugignore"; then
+  error << EOF
+'package.json' listed in '.slugignore' file
+
+The 'heroku/nodejs' buildpack is set on this application, but was
+unable to detect a 'package.json' file. This is likely because
+the '.slugignore' file is removing it before the build begins.
+
+For more information, refer to the following documentation:
+https://devcenter.heroku.com/articles/slug-compiler#ignoring-files-with-slugignore
+EOF
+elif [[ -f "$1/.gitignore" ]] && grep -Fxq "package.json" "$1/.gitignore"; then
+  error << EOF
+'package.json' listed in '.gitignore' file
```

Comment:
Would love to get numbers on this. Wonder of other languages might be doing something similar. Can you add an mcount?

Reply:
Unfortunately we don't have `bin/detect` connected to metrics, so we can't get numbers here yet :\